src/server/terrain/structure_building_blocks.zig

```diff
@@ -0,0 +1,149 @@
+const std = @import("std");
+
+const main = @import("root");
+const ZonElement = main.ZonElement;
+const Blueprint = main.blueprint.Blueprint;
+const List = main.List;
+const ListUnmanaged = main.ListUnmanaged;
+const AliasTable = main.utils.AliasTable;
+const Neighbor = main.chunk.Neighbor;
+const Block = main.blocks.Block;
+const Degrees = main.rotation.Degrees;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+var arena = main.heap.NeverFailingArenaAllocator.init(main.globalAllocator);
+const arenaAllocator = arena.allocator();
+
+var blueprintCache: std.StringHashMapUnmanaged(*[4]BlueprintEntry) = .{};
+var childrenToResolve: List(struct {parentId: []const u8, colorName: []const u8, colorIndex: usize, childIndex: usize, structureId: []const u8}) = undefined;
+
+const originBlockStringId = "cubyz:sbb/origin";
+var originBlockNumericId: u16 = 0;
+
+// Maps global child block numeric ID to index used to locally represent that child block.
+var childBlockNumericIdMap: std.AutoHashMapUnmanaged(u16, u16) = .{};
+var childBlockStringId: ListUnmanaged([]const u8) = .{};
+
+const BlueprintEntry = struct {
+	blueprint: Blueprint,
+	originBlock: StructureBlock,
+	childBlocks: []StructureBlock,
+
+	const StructureBlock = struct {
+		x: i32,
+		y: i32,
+		z: i32,
```

Comment:
Could these be `i16`? Blueprints shouldn't be that large anyways.

Reply:
Its a bit late for limiting capacity of blueprints, we already chose u32 by using Array3D, although 60k blocks is reasonable limitation it makes interface inconsistent, some parts support bigger blueprints, other wont.

But there is a bug here, those should be `u32` to match Array3D capacity limitations, not `i32`.

If you insist I can change it to u16, I don't think there is a use for that sign here.

Reply:
Oh, I thought you may have used the sign to store precomputed offsets from the origin.

Either way it should be 16 bits. The blueprint capturing just cannot reasonably capture such giant things anything, since the render distance (or actually I think it's limited by the entity animation distance even) doesn't get that high.

Reply:
Hmm I could be possible, but I am also not 100% sure I can do that, indexes don't like me, so I would prefer to tinker with it in #1227 to be able to see what are the results.

It will be u16 then.